environs/tools/tools.go

```diff
@@ -20,10 +20,23 @@ import (
 
 var logger = loggo.GetLogger("juju.environs.tools")
 
-func makeToolsConstraint(cloudSpec simplestreams.CloudSpec, stream string, majorVersion, minorVersion int,
+var streamFallbacks = map[string][]string{
+	ReleasedStream: {ReleasedStream},
+	ProposedStream: {ProposedStream, ReleasedStream},
+	DevelStream:    {DevelStream, ProposedStream, ReleasedStream},
+	TestingStream:  {TestingStream, DevelStream, ProposedStream, ReleasedStream},
+}
+
+func makeToolsConstraint(cloudSpec simplestreams.CloudSpec, stream string, fallback bool, majorVersion, minorVersion int,
 	filter coretools.Filter) (*ToolsConstraint, error) {
 
 	var toolsConstraint *ToolsConstraint
+
+	streams := []string{stream}
+	if fallback {
+		streams = streamFallbacks[stream]
```

Comment:
if they pass in a stream that's not in fallbacks, we need to just use that

Reply:
Ok, makes sense.